### DISCUSSION TITLE
Welcome to clifm Discussions!

### DISCUSSION THREAD

**Discussion by leo-arch** [leo-arch/clifm#1] [General]
## Welcome!
We’re using Discussions as a place to connect with people interested in this project. Here you can:

  * Ask whatever question you’re wondering about.
  * Share ideas and suggest new or improved features.

**Comment by tuerda** [comment]
Hello. I have been using this file manager quite heavily for a few days, and I would like to offer a very detailed review.

First and foremost: I love it. I probably had been looking for something like this for years, and didn't quite realize it.

In a sense, this feels almost like it could be a whole bunch of convenient shell aliases and such, but every time I try to think about how I would implement it in that way I realize there is more going on.
The improvements over just using the bare CLI to manage files are very subtle, but also highly significant.

It is so close to being a shell interface, though, that I thought for a bit about literally just using it as my shell.
I decided this probably was not a good idea, because it makes certain potentially dangerous operations too easy to do by mistake.

## Comments on the core functionality:

- Re-displaying the directory contents and file numbers is something that feels like it should have a more direct method.
There is a somewhat un-ergonomic keybinding, and there is also issuing the command `.` . The second method is the one I use, and it works fine, but I wonder, "why not do this if I just send a blank line as a command?".

- Symlinks in unix are a bit touchy. In the basic shell, if I do `ln -s file1 ../bin/stuff` it will actually do something really dumb, which is create a file in `../bin/stuff` named `file1` which does nothing at all because it is a symlink to itself.
In order to get the link to work, you either have to write up the relative path from the destination directory, or the absolute path for the file.
Which is preferred depends a little bit on the situation, but both are a bit of a hassle.
I was hoping that in clifm the `l` command would automatically solve this issue doing one of these approaches (the absolute path is probably the easiest to implement), but it did not. Using the filename itself I can kind of see why you might want to keep it faithful to the shell, but if I use the _number_ that corresponds to it, I clearly am talking about the file itself and not about the name as it appears in the list.

- Filters work backwards from what I would expect. If I filter for a regex, I would expect to see only files that match that regex, but instead I see only files that do _not_. This is not a big deal, since files that match a regex can still be shown using `/`, but I still find it mildly confusing, and wonder if maybe there should be a way to reverse the behavior of a filter.

## Bugs:

I used the hell out of this and got one bug in the core functions once. The text I typed into the commands was invisible. It was still working, and the output of commands displayed just fine, but the commands themselves did not. I restarted the program and tried to do the same things to see if I got the same result twice, but I did not. I am not sure what conditions triggered it.

## Plugins, extensions, etc:

While the basic functionality worked essentially without any problems at all, the same could not be said for the plugins that I tried.
I personally do not care about these plugins, and I tried them mostly to provide feedback.
I am also aware that these depend on some external programs and that some of this might involve the versions of the external programs which I am running, etc.

I tried and had trouble with two plugins: fzf-nav and icons.

#### icons:

This is the easiest one. I use a pretty unusual bdf font.
The icons-in-terminal project simply failed to patch in the icons, probably because it expected my font to be a more modern true type font like everyone else uses.
I installed, tried it in clifm, found that the icons were just squares, then went and tried the display icons script from the icons in terminal project and found that that displayed squares as well.
Hence, this is in no way your fault, but you might want to add a warning that it might not work with all fonts.

#### fzf-nav:

This mostly worked, but the issues were quite infuriating.
It is bad enough that I would never use this, _but_, I personally do not care:
The main attraction to it was the file previews, and I found that the fzf program simply was not the interface I would choose for file previews. I tried launching ranger nested inside clifm and it worked just fine, file previews and all, and the ranger interface was nicer than fzf anyway.

The issues were specifically with the BFG file previews. The parts that I looked into deeply were the display of images with ueberzug, and the display of pdf files with pdftotext.
I was unable to figure out how to change what program was used to display the previews of each filetype. In particular, pdf files were not being displayed with pdftotext, but instead the first page was displayed as an image with ueberzug. I tried to edit the fzf-nav script, as instructed, but nothing that I did changed this. I did not completely understand how the switches to test if a program was installed worked: I tested them separately in my shell and all of them were consistently evaluating to `true`, regardless of whether the program in question was installed or not.
I tried uninstalling ueberzug, and pdf files just gave a message that it was a binary pdf file, and did not display anything at all. I tried opting to use scope.sh instead, but then images and pdf files both just gave me an empty box (plain text files did display and used the usual scope.sh highlighting scheme).

Getting previews of pdf files as images is not necessarily a bad thing, mind you, but the way ueberzug was working _was_ pretty bad. Previews of images and pdfs were not displayed in their assigned box, but instead plastered over the directory view, covering up the rest of the interface and not allowing me to see what I was doing. It is this particular issue that - to my mind - renders the whole thing completely unusable.


## FINAL THOUGHTS:

Overall great work! This made enough of a difference in my workflow that I quickly cleaned up several messy parts of my filesystem that I had just allowed to fester with my previous tools. I am not sure how I ever lived without it!
There are some things that I think it might be possible to do better (symlinks in particular).
I also found that the plugins I tried had some rough edges.
Everything else is pretty much perfect though.

**Reply by leo-arch** [reply]
> Filters work backwards from what I would expect

The files filter works now as you said it should: `!REGEX` to exclude, and `REGEX` to include only those files matching the pattern.

**Reply by tuerda** [reply]
Just saw this now. Thanks for the update. This feels more intuitive to me.

**Comment by leo-arch** [comment]
> In a sense, this feels almost like it could be a whole bunch of convenient shell aliases and such, but every time I try to think about how I would implement it in that way I realize there is more going on.

Indeed, CliFM started as a simple attempt to create a bookmarks system for the shell. The first version(s) of the program was in fact a (horrible) [bash script](https://hithub.com/leo-arch/clifm-bash). But as soon as you try something a bit more complicated, the shell tools probe to be quite inefficient (desperately slow). A shell script was definitely not the adequate tool. This is why the whole thing was rewritten in C (I even tried Python, but... no). I you care about performance, C is no doubt the way to go.

> I thought for a bit about literally just using it as my shell

Exactly as you yourself figured it out, that's not a good idea at all. CliFM is a FILE MANAGER, tightly integrated to shell, true, but it is not a shell (I myself tried some time ago to convert the whole thing into a shell of its own, but realized that such an attempt was not only quite hard, but also unnecessary: we already have shells doing the same thing, and better for sure). However, I have not found any serious issue using CLiFM as a shell, at least running some basic shell commands: I myself do it all the time. But it is still NOT a shell.

> Re-displaying the directory contents and file numbers is something that feels like it should have a more direct method

`rf` (short for `refresh`) is a command specifically designed for this purpose. Not sure if it is as direct as you expect though. About the blank lines idea, I'm not sure about it: blank lines are mostly intended to refresh/update the prompt (and only the prompt). But I will consider it.

> Symlinks in unix are a bit touchy.

True. And it is also true that I tried to `keep it faithful to the shell`, just because I don't want to make decisions in place of users. However, sometimes it is needed (`c`, and `md` do this after all). But there would be not a problem at all to improve the `l` command. May you expand a bit more on what you exactly expect from this command: how should it behave exactly, under what conditions? Once again, the issue needs to be reproduced.

> Filters work backwards from what I would expect

Yes, by default **filters** exclude, and I just never thought about making it work the other way round. But it seems a quite reasonable demand, and quite easy to implement. It will be done.

About the **bug(s)**: No way to fix it if it cannot be reproduced.

**Plugins**:

Here you touch the two plugins I care the most, just because people really like this kind of stuff: I've seen people rejecting a file manager just because it cannot preview files. So, files previewing and icons are important.

This said, the **icons** issue is completely foreign: as you said, it depends on your fonts (though I didn't know/try it thus far) and on your terminal. I did was aware nonetheless of the terminal issue. `urxvt` do not correctly display (they look square) these icons (those provided by `icons-in-terminal` I mean). On most other terminals icons work just fine. I guess I should better document this feature. What I could try is to add support for other terminal icons projects, like nerd fonts. Do you know if nerd fonts work right on urxvt?

**Previewing** is a native feature of FZF, so I thought it would be a good tool for the job (and I should say it mostly works). I'm sure it is a matter of configuration. Just tell me exactly what terminal, fonts, and configuration are you using to improve the plugin. I need to reproduce the wrong behavior to fix it. It should not be any problem at all.

Needles to say, thank you very much for your time and report.

**Reply by tuerda** [reply]
The one I personally care about is the symlinks.
I want to be able to say `l 7 ~/some/directory/here` and have a new symlink in `~/some/directory/here` which points to the file that matches the ELN number 7.
If I wanted the twisted behavior of the `ln -s` command, I would use that (I am in the shell, after all).
My suggestion of how to do this would probably be to make the ELN numbers expand to absolute file paths rather than to just the names of the files; this sounds like it probably would not break anything else, and would make it so you didn't have to do anything special with `l`.

Icons:
I was pretty sure that the icons issue was because of the bitocra font rather than because of urxvt, but you might know better than I do.
I have never looked into nerdfonts etc.

Previewing:

You want to know about my setup, so here we go:

Ubuntu 18.04 (old, I know, but this is how it was provided by my workplace).
This actually has an outdated version of readline, so I had to compile clifm myself, which was no big deal.
I use the herbstluftwm window manager, the urxvt terminal and the bitocra bdf font.
I am running zsh, but not any of those "oh my zsh" style framework things.
For fzf, I downloaded the latest binary from the github releases page.
I downloaded only the executable program and none of the extras, and I put a symlink to this program in my path and that seemed to be enough for clifm.

I don't like fzf very much, which I realize makes me a total weirdo, but it is also part of the reason why I don't care about this. I think you might care though, because clifm seems to be made specifically for people like myself with highly customized setups.

Much love.

**Reply by leo-arch** [reply]
Great. So, I will fix the symlinks issue, try bitocra fonts, and reproduce (I hope I can), the previewing problem.

**Reply by tuerda** [reply]
The previewing problem is 100% reproducible for me. It happens every single time. I can record a video of it, send you any relevant files, command outputs, etc. Just tell me what you need.

**Reply by leo-arch** [reply]
Well, the symlinks issue is fixed. Try it an let me know.

**Reply by tuerda** [reply]
Getting some compiler errors now. This is probably because of Ubuntu 18.04 having very outdated libraries. I will try and debug it tomorrow. (maybe I can just use the binary file and simlink the readline 7 libs to readline 8?: It compiled fine with readline 7 before . . . )

**Reply by leo-arch** [reply]
Weird. However, it is generally not a good idea to symlink libraries that way. But try it.
I you have any problem send me the compiler logs, maybe I can help you.

**Reply by leo-arch** [reply]
Trying now the bitocra fonts. So tiny! Other than that the file previewer still works fine on urxvt using these fonts.

**Reply by tuerda** [reply]
Hello again!

I had to install some libraries which were apparently not required 3 days ago (?) but I installed them, got it compiled, and now . . . it is constantly segfaulting. My money is very much on the outdated ubuntu libraries. This program is useful enough that I will insist on getting it working, but it might take a while.

**Reply by leo-arch** [reply]
Bad to hear that. Ubuntu is... well.

**Reply by tuerda** [reply]
OK, not so hard after all. Just completely wiped all of the intermediate build files and compiled again from scratch: Now it works and symlinks behave as I hoped.

Thank you!

The issues with the previewer I cannot explain: I mean, fzf was straight from github and überzeug was from pip. The rest of it seems to just be a bash script, so I really can't see where outdated software could have been at fault.
I tried again and the images are definitely still not going in their designated box, and still covering up the directory view, so I got nothing.

As I said previously, I personally do not care, but you might, and I am willing to help in any way that seems appropriate.

**Reply by leo-arch** [reply]
Great! I installed Ubuntu 18.04 on a VM just to reproduce the issue. No luck yet. It isn't perfect for sure (gifs previewing is broken), but it works.
And it seems you're right: the previewing plugin is a mess. Working on it.

**Reply by leo-arch** [reply]
And please, keep in touch. You've been of big help. Just let me know whatever doesn't work, or needs to be somehow modified, for whatever reason. Source code is as important as users feedback. Thank you very much @tuerda!

**Reply by leo-arch** [reply]
Btw, have you tried the .deb package in the [Releases](https://github.com/leo-arch/clifm/releases/tag/v1.1) page? I've tried it on my own Debian box, and it works, but who knows.
Bear in mind however that the packaged release is not fully up to date: the symlinks fix is not there yet.

**Reply by tuerda** [reply]
The deb package was the first thing I tried. It dynamically links to readline 8 and ubuntu 18.04 only has readline 7. This is why I ended up compiling myself, and also what I meant previously about making a symlink, (but I did not end up doing that).

**Reply by tuerda** [reply]
Hey there again, something is still not working quite right with symlinks.

`l ~/some/directory/some/file ./`
`ls -lh file`

and we get
```
metainfowedon'tcareabout             file -> /where/I/am/now/~/some/directory/some/file
```
which is a nonsense path to link to.
It seems like this is always assuming that the file we are making the symlink to lives in the current directory.
That's the only way I had tried it so far, which is why I didn't notice earlier.

EDIT: Thinking about this some more: Stupid unix symlinks. I think it might be solvable just by checking if the target of the symlink starts with `~`, `/` or `.` . . . maybe? . . . there might be some case I'm not thinking of . . . directory aliases perhaps?

**Reply by leo-arch** [reply]
Absolutely, you're right. I checked only for `/`. Just forgot to check for tilde and dot. It will be fixed tomorrow.

**Reply by leo-arch** [reply]
Ok, couldn't wait. Should be fixed now.

**Reply by leo-arch** [reply]
Btw, you can use the `p` command to get files properties (including symlinks target) instead of using `ls -lh`.

**Reply by tuerda** [reply]
I didn't actually write a reply, but just gave a thumbs up. I thought you would read that as "it seems to be working as expected now", but maybe it was not clear. Hence, I write this now to say so explicitly: Looks like it is all in order :)

**Reply by leo-arch** [reply]
The thumbs up were fine. Don't worry.

**Reply by leo-arch** [reply]
Still trying to reproduce the previewer issue: no success. This is an Ubuntu 18.04 box, plain zsh, urxvt, and bitocra fonts.

![ubuntu](https://user-images.githubusercontent.com/21303106/120706960-1dd89500-c490-11eb-819d-edb884329ac2.png)

Herbstluftwm perhaps?

However, it was useful to try this setup to fix of couple of minor things. So, not entirely unsuccessful.

**Reply by tuerda** [reply]
Hey, hadn't checked here in a while. I booted up into the default gnome dektop just to check. Here is a screenshot:
![2021-06-03-202012_1292x731_scrot](https://user-images.githubusercontent.com/38886697/120727927-bfb8ab80-c4a9-11eb-94b7-f1e17b128c20.png)
As you can see, the image is still in the wrong place. You can't even tell what is going on, really, because it is covering up the whole interface. You can only see the top and the (empty) frame on the right side.

It may be that this problem is something really specific to something in my setup that others do not have. If this is the case, then don't worry about it because I don't use this.

**Reply by leo-arch** [reply]
It really bothers me not to be able to fix this issue. I'm sure I will eventually find it and fix it.

For what I can see, two values are wrong: image size and horizontal position (X). These values are defined in `BFG.sh`, lines 34-35 (WIDTH and HEIGHT, both fixed values), and 49 (X). Perhaps the problem is there, though I don't know why, because these settings work on most other setups.

Btw, nice drawing!

**Reply by tuerda** [reply]
I am willing to try any kind of random experiment you like. Just let me know what to try.

BTW: Haha, thanks. It is a pencil drawing of my girfriend sleeping; it was the first thing in my images directory. Every time I look at it I think the thumb is too small . . .

**Reply by leo-arch** [reply]
Perhaps the thumb is a bit small, but the perspective effect is amazing, and the sheets! Great job! It's a bit like CliFM: it overalls works great, but this previewer issue really annoys me, just the that thumb annoys you, because you are the author (for other people though is just a very minor thing).

Here's what I would like to try:

File: `BFG.sh`

Lines 34-5:
```sh
# Default size for images
#WIDTH=1920
#HEIGHT=1080
WIDTH=400
HEIGHT=400
```

Lines 43-51 (calculate_position function)
```sh
calculate_position() {
	# shellcheck disable=SC2034
	read -r TERM_LINES TERM_COLS << EOF
	$(</dev/tty stty size)
EOF

	# X=$((TERM_COLS - COLUMNS - 2))
        X=50
	 Y=2
}
```
These edits should produce some modification in image previews, so that we know thus that the problem is actually there. I can send you the modified files if you want.

**Reply by leo-arch** [reply]
Before trying the new settings, delete the cache directory `($HOME/.cache/clifm/previews`). Otherwise, the previewer will keep loading cached images. This is why it kept showing PDF images instead of text, despite the fact that you switched from `pdftoppm` to `pdftotext`. I need to rework all this.

**Reply by tuerda** [reply]
OK, had been busy all day. Tried now.

The effect is that images are now no longer visible at all, anywhere. I thought maybe they were going someplace off screen, but I scaled up my terminal to be _much_ wider than 50 columns and still there was no visible effect.

EDIT: BTW, this has got to be the simplest plugin ever but I adapted my slmenu directory switching script to work in clifm:

```
#!/bin/bash
printf "$(find -type d 2> /dev/null | slmenu -i -l 20)" > "$CLIFM_BUS"
exit 0
```
I am probably the only person on earth who uses slmenu. You may change `slmenu -i -l 20` it to `fzf`.

**Reply by leo-arch** [reply]
> The effect is that images are now no longer visible at all

Well, at least the plugin is usable now :smile: . Maybe is off the screen, yes. I'll experiment a bit more with those values.

Didn't even know about `slmenu`. The good thing is that it works. With just one line of code you built your own plugin. Great.
Btw, `fzfnav` is also (indeed, it is fundamentally) a directory switcher: enter the directory you want, exit pressing <kbd>q</kbd> and CliFM will switch to that directory.

**Reply by tuerda** [reply]
I tried to restore the position (lines 45-51) and change only the size:
```
WIDTH=10
HEIGHT=10
```
This put me right back where we were before. The images are back, in the wrong place, and way too big (even with this _drastic_ change in default sizing).

I found another place which seemed to have some hardcoding of width and height variables on line 361. I tried to make those 50x20 instead of 500x290 to see if that did anything; it did not.

----

Going off topic slightly:

`slmenu` is a very old program which was maintained for all of about a month circa 2013 or so.
I am no longer able to find it online anywhere, but I have been keeping my own backup around since then.
You could think of it as a drop-in replacement for fzf in any script that doesn't require the previewer.
It is literally just taking the dmenu source code and making the smallest change necessary for it run in a terminal instead of as a GUI.
There are two things I like about it: It draws the menu below the prompt without redrawing the terminal screen so I can still see output of previous commands. It also is not fuzzy, so I don't get tons of false positives.

The one line plugin is really just taking `` alias dcd='cd "$(find -type d 2> /dev/null | slmenu -i -l 20 )" ' `` and convincing it to work in clifm. It is a little more like `jumper` than like `fzfnav` . . . the point is just to go someplace directly without any of the intermediate steps. If you add your usual boilerplate "check if fzf is installed . . . bla bla" it could be a reasonable alternative.

**Reply by leo-arch** [reply]
Don't worry about the previewer issue: it will eventually be solved. For the time being, specially because it seems to happen only in your specific setup, we will let it live. However, I'm reworking the previewer with a few important changes:

a) Use hashes (instead of file names) for cached images: many files could have the same name, and that's a problem.
b) Add a configuration file (BFG.cfg) for easy setup. It is not that configurable right now, but at least lets you choose which of the available applications to use for different file types: asking to user to edit the code is definitely not a friendly gesture.

As to `slmenu`, the problem is that it cannot be found anywhere, and thereby adding it as an alternative to fzf makes not much sense. As soon as I realized that it was not in Arch repos (not even in the AUR!), I knew it was a very rare program. I guess it is something like `dmenu`, isn't it?

**Reply by tuerda** [reply]
This sounds like cool stuff, although I don't really need a previewer so I probably will not use it (except for testing, if you ask me to).

Yes, as I said, slmenu is off-topic.
I think I might be the only person on earth who uses slmenu (I use the hell out of it though). I did not mean to suggest you should offer slmenu as an alternative to fzf. I write and maintain my own slmenu scripts, and I do not expect you to participate.

I would also be willing to put slmenu on github if anyone but me showed signficant interest. As far as I can tell, however, nobody cares, so I just keep my own backup.

**Reply by leo-arch** [reply]
A Github repo would be a nice idea. Maybe people don't care just because they don't know, not because `slmenu` isn't worthwhile. I would definitely give it a try.

Btw, could you please take a look at [this](https://github.com/leo-arch/clifm/discussions/19)?

**Reply by leo-arch** [reply]
> a) Use hashes (instead of file names) for cached images: many files could have the same name, and that's a problem.
b) Add a configuration file (BFG.cfg) for easy setup.

Done. though the previewer issue is still there, `fzfnav` should be better now: most features can be disabled if you don't like/need them, and previewing applications can be chosen from a (limited) list of supported applications.

**Comment by leo-arch** [comment]
I almost forgot it: I will improve fzfnav and BFG scripts to make them more easily hackable. Better documentation for this is needed too. Consider most plugins as beta features. Btw, do you know of any file previewer better than FZF?

**Reply by tuerda** [reply]
If we are talking about previewing programs for the terminal, I honestly feel like ranger knocks this out of the park.

I actually kind of like ranger, and I would happily use it if TUI was my jam. I used to think that it was, but clifm has pretty much put the final nail on that coffin.

**Comment by leo-arch** [comment]
Regarding the re-displaying thing, have you tried rebinding the keyboard shortcut to something more ergonomic? You could also wright a simple alias for the `rf` command, I mean, some single char like z or something.

**Reply by tuerda** [reply]
I am using `.` for this purpose.
It is pretty direct and works fine.

This is more like an "If I had written this, the way I would have done it is . . . " rather than an actual criticism. I just feel like the list is very important in clifm and even might have thought to automatically re-display if it ever scrolls off screen (this is too much, though: You might cover up some important info)

**Reply by leo-arch** [reply]
> re-display if it ever scrolls off screen

Do you mean something like repeatedly pressing Enter in the prompt till the list just disappears? In that case, it shouldn't be hard to implement. Maybe as an optional feature.

**Reply by tuerda** [reply]
That is an option I guess.
TBH I don't care about this very much and I don't think you should lose sleep over it either: If I had made clifm it would have been very eager to show you this list all the time.

The way you did it is different, but it is sensible and good.

**Reply by leo-arch** [reply]
I'll bear you suggestion in mind anyway.
